### PR TITLE
add restore to alternate resource

### PR DESCRIPTION
**Pull request by ryanfkeepers**
adds support for restoring to a resource that
differs from the one whose data appears in the backup.

---

#### Does this PR need a docs update or release note?

- [x] :clock1: Yes, but in a later PR

#### Type of change

- [x] :sunflower: Feature

#### Issue(s)

* #3562

#### Test Plan

- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

File: src/internal/m365/onedrive/restore_caches.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
+package onedrive
```
Comment (ryanfkeepers) [inline src/internal/m365/onedrive/restore_caches.go:1]: This file is purely lift-and-shift. it was moved from `restore.go` since it has grown from a small container of maps to a struct with constructors, populators, and other methods.  No changes to this code have been made in this PR outside of the movement.

File: src/internal/m365/exchange/restore.go [hunk 2]
```diff
@@ -40,16 +40,13 @@ func ConsumeRestoreCollections(
...
 	var (
-		userID         = dcs[0].FullPath().ResourceOwner()
+		userID         = rcc.ProtectedResource.ID()
 		directoryCache = make(map[path.CategoryType]graph.ContainerResolver)
 		handlers       = restoreHandlers(ac)
 		metrics        support.CollectionMetrics
 		el             = errs.Local()
 	)
 
-	// FIXME: should be user name
```
Comment (pandeyabs) [inline src/internal/m365/exchange/restore.go:50]: Curious, why was this removed?

File: src/internal/m365/exchange/restore.go [hunk 3]
```diff
@@ -40,16 +40,13 @@ func ConsumeRestoreCollections(
 	}
 
 	var (
-		userID         = dcs[0].FullPath().ResourceOwner()
+		userID         = rcc.ProtectedResource.ID()
```
Comment (pandeyabs) [inline src/internal/m365/exchange/restore.go:43]: nit: Call it `protectedResourceID` instead of `userID`?

**Review by pandeyabs — APPROVED** [review]

File: src/internal/m365/exchange/restore.go (hunk 2 above)
Comment (ryanfkeepers) [reply src/internal/m365/exchange/restore.go:50]: It's a duplicate, and is already present in clues from a parent caller.

File: src/internal/m365/exchange/restore.go (hunk 3 above)
Comment (ryanfkeepers) [reply src/internal/m365/exchange/restore.go]: I'll go with `resourceID` for a middle ground.
